### PR TITLE
Feature: support speed test of URLTest through API

### PR DESCRIPTION
**Pull request by tslling**
support manually triggering a speed test of URLTest proxy group through API. #273

### CODE REVIEW

File: hub/route/proxies.go [hunk 1]
```diff
@@ -135,3 +136,14 @@ func getProxyDelay(w http.ResponseWriter, r *http.Request) {
...
+
+func reelectProxy(w http.ResponseWriter, r *http.Request) {
+	proxy := r.Context().Value(CtxKeyProxy).(*A.Proxy)
+	urlTest, ok := proxy.ProxyAdapter.(*A.URLTest)
+	if !ok {
+		render.Status(r, http.StatusBadRequest)
+		render.JSON(w, r, newError("Must be a Url-Test"))
+		return
+	}
+	urlTest.SpeedTest()
```
Comment (popu125) [inline hub/route/proxies.go]: Maybe it need a response.

File: adapters/outbound/urltest.go [hunk 2]
```diff
@@ -71,6 +71,10 @@ func (u *URLTest) Destroy() {
 	u.done <- struct{}{}
 }
 
+func (u *URLTest) SpeedTest() {
+	u.speedTest()
```
Comment (popu125) [inline adapters/outbound/urltest.go]: I think its better to do speedtest concurrently.

The implement of `speedTest()` does prevent race condition, it's safe to just add a `go`.

**Comment by popu125** [comment]
个人对 API 名字的选择持保留态度（虽然并没有什么卵用

File: hub/route/proxies.go (hunk 1 above)
Comment (tslling) [reply hub/route/proxies.go]: You mean response before speedtest? By doing this, we return nothing useful to client, how about waiting until the speedtest finished, then we can return the picked proxy to client. But obviously we can not wait all the time, maybe client will supply a ```timeout```.

**Pull request closed without merging**